15840 - Ensures all Discovery fetch gets called and abort ongoing ones.

JoeyAndres opened the pull request:
- Discovery doesn't fetch subsequent requests when there's an
  on-going fetch.
- This fixes that by queueing the data needed by SavedSearchSource
  during fetch and dequeuing them once the ongoing request
  is done or aborted.

This is my preliminary proposed solution to the [problem](https://github.com/elastic/kibana/issues/15840) I posted. I encourage you to reproduce the problem and play with this. 

I'm hired to work by on company's Kibana instance and I want to have this opportunity to participate and contribute to this project long term. Please give me feedback on all things with respect to my solution to this issue. From code styles, tests (will add them), better modules to modify, etc...

elasticmachine commented:
Can one of the admins verify this patch?

Bargs commented:
Thanks for the contribution @JoeyAndres, I'll try to take a look at this as soon as I can.

JoeyAndres commented:
@Bargs As I run more mileage on this test, I added some important changes.
- Instead of aborting on-going requests, skip them.
- Using the abort promise assumed that no one else calls it, but it looks like it is being resolved when we transition to some other page other than discovery.
  - This causes the Promise loop in search_source.js (which listens to both complete and abort promise) to resolve, creating a new requests when we exit of discovery page
  - New calls to fetch at this point will call these hanging requests queued during the exit of discovery page.
  - Causing 'divide by zero' errors since the time range doesn't exist anymore in the $state.
- Adding a **skip** promise in AbstractRequest fixes this.

**Problem:** $scope.fetchQueue's requests will all be called to finished, instead of aborting ongoing ones.

Probably will not concern you all. But in our organization, I created an internal plugin for Kibana and we modified elasticsearch so that they do the following:

1. When in discovery page, append 'kibana-page-id-session' with a unique id generated during kibana is first loaded.
  * This ensures we have a unique id for each kibana page/tab/window.
2. In elasticsearch, if 'kibana-page-id-session' is present in the header of an msearch, abort the last one with similar 'kibana-page-id-session' and send an empty results for those.
  * This have an observed effect of faster query like abort.
  * Due to the id being unique per page, we don't abort msearch from other page/tab/window.
  * Since elasticsearch abort is only done when page-id-session, other msearch usage (e.g. console or whatever) should be safe.

joeyJsonar commented:
A more up to date fix: https://github.com/elastic/kibana/pull/16218
